prov/rxd/src/rxd_ep.c

```diff
@@ -0,0 +1,1789 @@
+	}
+
+	fastlock_acquire(&ep->domain->lock);
+	dlist_insert_tail(&ep->dom_entry, &ep->domain->ep_list);
+	fastlock_release(&ep->domain->lock);
+	ret = 0;
+out:
+	rxd_ep_unlock_if_required(ep);
+	return ret;
+}
+
+struct rxd_peer *rxd_ep_getpeer_info(struct rxd_ep *ep, fi_addr_t addr)
+{
+	return &ep->peer_info[addr];
+}
+
+void rxd_ep_lock_if_required(struct rxd_ep *ep)
+{
+	/* todo: do locking based on threading model */
+	fastlock_acquire(&ep->lock);
+}
+
+void rxd_ep_unlock_if_required(struct rxd_ep *ep)
+{
+	/* todo: do unlocking based on threading model */
+	fastlock_release(&ep->lock);
+}
+
+size_t rxd_get_msg_len(const struct iovec *iov, size_t iov_count)
+{
+	size_t i, ret = 0;
+	for (i = 0; i < iov_count; i++)
+		ret += iov[i].iov_len;
+	return ret;
+}
+
+static void rxd_init_ctrl_hdr(struct ofi_ctrl_hdr *ctrl,
+			      uint8_t type, uint16_t seg_size,
+			      uint32_t seg_no, uint64_t msg_id,
+			      uint64_t rx_key, uint64_t source)
+{
+	ctrl->version = OFI_CTRL_VERSION;
+	ctrl->type = type;
+	ctrl->seg_size = seg_size;
+	ctrl->seg_no = seg_no;
+	ctrl->msg_id = msg_id;
+	ctrl->rx_key = rx_key;
+	ctrl->conn_id = source;
+}
+
+static void rxd_init_op_hdr(struct ofi_op_hdr *op, uint64_t data,
+			    uint64_t msg_sz, uint8_t rx_index,
+			    uint8_t op_type, uint64_t tag, uint32_t flags)
+{
+	op->version = OFI_OP_VERSION;
+	op->rx_index = rx_index;
+	op->op = op_type;
+	op->op_data = 0; /* unused */
+	op->flags = flags;
+	op->size = msg_sz;
+	op->data = data;
+	op->tag = tag;
+}
+
+static uint32_t rxd_prepare_tx_flags(uint64_t flags)
+{
+	uint32_t tx_flags = 0;
+	if (flags & FI_REMOTE_CQ_DATA)
+		tx_flags = OFI_REMOTE_CQ_DATA;
+	if (flags & FI_TRANSMIT_COMPLETE)
+		tx_flags |= OFI_TRANSMIT_COMPLETE;
+	if (flags & FI_DELIVERY_COMPLETE)
+		tx_flags |= OFI_DELIVERY_COMPLETE;
+	return tx_flags;
+}
+
+uint64_t rxd_copy_from_iov(const struct iovec *iov, size_t iov_count,
+				 void *dst, uint64_t data_sz, uint64_t skip)
+{
+	int i;
+	uint64_t rem, offset, len, iov_offset;
+	offset = 0, rem = data_sz, iov_offset = 0;
+
+	for (i = 0; i < iov_count; i++) {
+		len = iov[i].iov_len;
+		iov_offset = 0;
+
+		if (skip) {
+			iov_offset = MIN(skip, len);
+			skip -= iov_offset;
+			len -= iov_offset;
+		}
+
+		len = MIN(rem, len);
+		memcpy((char *)dst + offset,
+		       (char *)iov[i].iov_base + iov_offset, len);
+		rem -= len, offset += len;
+	}
+	return offset;
+}
+
+uint64_t rxd_ep_copy_data_to_iov(const void *src, uint64_t data_sz,
+			       struct iovec *iov, size_t iov_count, uint64_t skip)
+{
+	int i;
+	uint64_t rem, offset, len, iov_offset;
+	offset = 0, rem = data_sz, iov_offset = 0;
+
+	for (i = 0; i < iov_count; i++) {
+		len = iov[i].iov_len;
+		iov_offset = 0;
+
+		if (skip) {
+			iov_offset = MIN(skip, len);
+			skip -= iov_offset;
+			len -= iov_offset;
+		}
+
+		len = MIN(rem, len);
+		memcpy((char *) iov[i].iov_base + iov_offset,
+		       (char *) src + offset, len);
+		rem -= len, offset += len;
+	}
+	return offset;
+}
```

Comment:
The above two functions are quite similar. Can they be merged or have the common portion in a separate function?